File: src/util/keychain.py
"""macOS Keychain 조회 wrapper + KIS(한국투자증권)/Telegram 키 자동 inject."""
from __future__ import annotations

import logging
import os
import subprocess

log = logging.getLogger(__name__)


def keychain_get(service: str, account: str) -> str | None:
    """평문 값 조회. SSH non-interactive에선 ACL로 차단될 수 있음 (GUI/launchd OK)."""
    try:
        r = subprocess.run(
            ["security", "find-generic-password", "-s", service, "-a", account, "-w"],
            capture_output=True, text=True, timeout=5,
        )
        if r.returncode == 0:
            return r.stdout.strip()
        return None
    except Exception as e:
        log.warning("keychain access 실패 (%s/%s): %s", service, account, e)
        return None


def load_kis_keys(mode: str | None = None) -> dict:
    """
    Keychain에서 KIS 키 → os.environ 자동 inject.

    Keychain service 이름 'kis-openapi' (2026-07-06 kiwoom-openapi에서 개명 —
    브로커는 처음부터 한투 KIS였고 kiwoom은 초기 kiwoom-mcp 시절 네이밍 잔재).
    account 10자리 = 앞 8자리(CANO) + 뒤 2자리(ACNT_PRDT_CD) 자동 분리.
    """
    mode = (mode or os.environ.get("KIS_MODE") or "paper").lower()
    loaded: dict = {}

    # app key / secret
    for env_var, ks_a in [
        ("KIS_APP_KEY",    f"{mode}-appkey"),
        ("KIS_APP_SECRET", f"{mode}-secret"),
    ]:
        if os.environ.get(env_var):
            loaded[env_var] = "already-set"
            continue
        v = keychain_get("kis-openapi", ks_a)
        if v:
            os.environ[env_var] = v
            loaded[env_var] = f"keychain ({len(v)} chars)"
        else:
            loaded[env_var] = "missing"

    # account 10자리 → 8 (CANO) + 2 (ACNT_PRDT_CD) 분리
    if not (os.environ.get("KIS_CANO") and os.environ.get("KIS_ACNT_PRDT_CD")):
        acct_full = keychain_get("kis-openapi", f"{mode}-account") or ""
        if len(acct_full) == 10:
            os.environ["KIS_CANO"] = acct_full[:8]
            os.environ["KIS_ACNT_PRDT_CD"] = acct_full[8:]
            loaded["KIS_CANO"] = f"keychain (8 chars, split from {len(acct_full)})"
            loaded["KIS_ACNT_PRDT_CD"] = f"keychain (2 chars, split)"
        elif acct_full:
            os.environ["KIS_CANO"] = acct_full
            loaded["KIS_CANO"] = f"keychain ({len(acct_full)} chars — not 10!)"
            loaded["KIS_ACNT_PRDT_CD"] = "default 01"
            os.environ.setdefault("KIS_ACNT_PRDT_CD", "01")
        else:
            loaded["KIS_CANO"] = "missing"
            loaded["KIS_ACNT_PRDT_CD"] = "missing"

    os.environ.setdefault("KIS_MODE", mode)
    return loaded


def load_telegram_keys() -> dict:
    """Keychain에서 Telegram 토큰 → os.environ 자동 inject."""
    loaded: dict = {}
    mapping = {
        "TELEGRAM_BOT_TOKEN": ("telegram-bot", "stock-trader"),
        "TELEGRAM_CHAT_ID":   ("telegram-bot", "stock-trader-chatid"),
    }
    for env_var, (svc, acct) in mapping.items():
        if os.environ.get(env_var):
            loaded[env_var] = "already-set"
            continue
        v = keychain_get(svc, acct)
        if v:
            os.environ[env_var] = v
            loaded[env_var] = f"keychain ({len(v)} chars)"
        else:
            loaded[env_var] = "missing"
    return loaded

File: cli/keychain.ts
/**
 * macOS Keychain read/write wrapper.
 *
 * Writing is the whole reason this module exists. `security`'s own usage text
 * says "Use of the -p or -w options is insecure" — the password lands in the
 * process's argv, which every user on the box can read out of `ps`. So the
 * write path never puts the secret in argv at all: it hands the *entire*
 * command line to `security -i` on stdin, which reads the same syntax from
 * standard input.
 *
 * Reading takes no secret argument, so `find-generic-password -w` is safe as
 * argv.
 *
 * Both paths run non-interactively. A `security` call against a locked login
 * keychain does not fail generically — it returns `-25308 User interaction is
 * not allowed`, the credential prompt that can only be answered in a GUI
 * session. That case gets its own error type so callers can tell the user to
 * unlock rather than reporting an opaque failure.
 *
 * The service/account names below are the wire contract with
 * `src/util/keychain.py`, which reads these same items at runtime. They must
 * match it exactly.
 */

import { execFileSync } from "node:child_process";

export const KIS_SERVICE = "kis-openapi";
export const TELEGRAM_SERVICE = "telegram-bot";
export const TELEGRAM_TOKEN_ACCOUNT = "stock-trader";
export const TELEGRAM_CHATID_ACCOUNT = "stock-trader-chatid";

/**
 * The signal pipeline's own credentials, read by `load_signal_keys()` in
 * `src/util/keychain.py`. The Python side looks each item up by these exact
 * strings, so a rename on either side is a silent runtime "missing", never a
 * compile error — `cli/__tests__/setup.test.ts` pins both halves together.
 */
export const SIGNAL_SERVICE = "signal-bot";
export const KRX_ID_ACCOUNT = "krx-id";
export const KRX_PW_ACCOUNT = "krx-pw";
export const BRAVE_KEY_ACCOUNT = "brave-api-key";

/** `security` is bounded so a blocked read fails the command instead of hanging a job. */
const TIMEOUT_MS = 10_000;

export function kisAccount(
  mode: "paper" | "real",
  kind: "appkey" | "secret" | "account",
): string {
  return `${mode}-${kind}`;
}

export class KeychainLockedError extends Error {
  constructor() {
    super(
      "macOS keychain is locked or this session cannot prompt (-25308). " +
        "Unlock the login keychain in a Terminal window and re-run.",
    );
    this.name = "KeychainLockedError";
  }
}

/** Feeds the command line to `security -i` on stdin; fd 0 is a pipe, never a terminal. */
type RunWrite = (args: string[], input: string) => string;

/** Reading needs no input, so fd 0 is detached outright. */
type RunRead = (args: string[]) => string;

const defaultRun: RunWrite = (args, input) =>
  execFileSync("security", args, {
    input,
    encoding: "utf8",
    timeout: TIMEOUT_MS,
    stdio: ["pipe", "pipe", "pipe"],
  });

const defaultGet: RunRead = (args) =>
  execFileSync("security", args, {
    encoding: "utf8",
    timeout: TIMEOUT_MS,
    stdio: ["ignore", "pipe", "pipe"],
  });

function text(value: unknown): string {
  return value === undefined || value === null ? "" : String(value);
}

/** Everything `security` printed, which is where the failure code shows up. */
function combinedOutput(err: unknown): string {
  const e = err as { stdout?: unknown; stderr?: unknown } | null;
  return `${text(e?.stdout)}\n${text(e?.stderr)}`;
}

function isLocked(err: unknown): boolean {
  const out = combinedOutput(err);
  return out.includes("-25308") || out.includes("User interaction is not allowed");
}

function firstLine(err: unknown): string {
  const e = err as { stderr?: unknown } | null;
  const stderr = text(e?.stderr).trim();
  const raw = stderr !== "" ? stderr : err instanceof Error ? err.message : text(err);
  return raw.split("\n")[0] ?? "";
}

/**
 * The names are interpolated into the stdin command line, where whitespace
 * separates arguments and quotes group them — a name carrying either would
 * change the command `security` ends up running.
 */
function assertName(service: string, account: string): void {
  if (/[\s"']/.test(service) || /[\s"']/.test(account)) {
    throw new Error("invalid keychain service/account name");
  }
}

/**
 * Store `secret` under `service`/`account`, replacing any existing item (`-U`).
 *
 * The secret reaches `security` only over stdin and never appears in a thrown
 * message — including when `security` echoes it back in its own diagnostics.
 */
export function keychainSet(
  service: string,
  account: string,
  secret: string,
  run: RunWrite = defaultRun,
): void {
  assertName(service, account);
  // A newline would terminate the command line and turn the rest of the secret
  // into a second `security` command. Checked before the general whitespace
  // rule below so this specific cause reports its specific remedy.
  if (secret.includes("\n")) {
    throw new Error("secret must not contain a newline");
  }
  // Measured, not theoretical: `security -i` splits the command line on
  // whitespace, so `-w has space` parses as `-w has` plus a positional
  // argument that add-generic-password reads as the *keychain name*. On an
  // unlocked keychain that stores the truncated value `has` and exits 0 — a
  // silently corrupted credential that later surfaces only as an auth
  // failure. Every credential this module stores (KIS app key/secret,
  // 10-digit account, Telegram token/chat id) is base64(url) or digits, so
  // none can legitimately carry whitespace or a quote: reject loudly.
  if (/[\s"']/.test(secret)) {
    throw new Error("secret must not contain whitespace or a quote");
  }

  const command = `add-generic-password -U -s ${service} -a ${account} -w ${secret}\n`;
  try {
    run(["-i"], command);
  } catch (err) {
    if (isLocked(err)) throw new KeychainLockedError();
    const detail = secret === "" ? firstLine(err) : firstLine(err).replaceAll(secret, "***");
    throw new Error(`keychain write failed for ${service}/${account}: ${detail}`);
  }
}

/**
 * Read the item back, or `null` when it does not exist.
 *
 * Absent and locked are deliberately different outcomes: a locked keychain
 * reported as "not configured" would send the user off to re-enter
 * credentials that are already stored.
 */
export function keychainGet(
  service: string,
  account: string,
  run: RunRead = defaultGet,
): string | null {
  try {
    return run(["find-generic-password", "-s", service, "-a", account, "-w"]).trim();
  } catch (err) {
    if (isLocked(err)) throw new KeychainLockedError();
    return null;
  }
}

export function keychainHas(
  service: string,
  account: string,
  run: RunRead = defaultGet,
): boolean {
  return keychainGet(service, account, run) !== null;
}
